libsoftwarecontainer/src/container.cpp

```diff
@@ -574,62 +574,92 @@ ReturnCode Container::destroy(unsigned int timeout)
-        log_error() << "Could not bind mount " << pathOnHost << " to " << dst;
+    if (isError(touch(tempDir))) {
+        log_error() << "Could not create file " << tempDir;
         return ReturnCode::FAILURE;
     }
+    m_cleanupHandlers.push(new FileCleanUpHandler(tempDir));
 
-    result = gatewaysDirInContainer() + "/" + pathInContainer;
-    if (readonly) {
-        return remountReadOnlyInContainer(result);
-    }
-
-    return ReturnCode::SUCCESS;
+    return bindMountCore(pathOnHost, pathInContainer, tempDir, readonly);
 }
 
 ReturnCode Container::bindMountFolderInContainer(const std::string &pathOnHost,
                                                  const std::string &pathInContainer,
-                                                 std::string &result,
                                                  bool readonly)
 {
-    if(isError(ensureContainerRunning())) {
+    if (isError(ensureContainerRunning())) {
         log_error() << "Container is not running or in bad state, can't bind-mount folder";
         return ReturnCode::FAILURE;
     }
 
-    std::string dst = gatewaysDir() + "/" + pathInContainer;
+    std::string tempDir = gatewaysDir() + "/" + std::string(basename(pathInContainer.c_str()));
+
+    log_debug() << "Creating folder : " << tempDir;
+    if (isError(createDirectory(tempDir))) {
+        log_error() << "Could not create folder " << tempDir;
+        return ReturnCode::FAILURE;
+    }
+
+    return bindMountCore(pathOnHost, pathInContainer, tempDir, readonly);
+
+}
 
-    log_debug() << "Creating folder : " << dst;
-    if (isError(createDirectory(dst))) {
-        log_error() << "Could not create folder " << dst;
+ReturnCode Container::bindMountCore(const std::string &pathOnHost,
+                                    const std::string &pathInContainer,
+                                    const std::string &tempDir,
+                                    bool readonly)
+{
+    if (pathInContainer.front() != '/') {
+        log_error() << "Provided path '" << pathInContainer<< "' is not absolute!";
         return ReturnCode::FAILURE;
     }
 
-    if (isError(bindMount(pathOnHost, dst, readonly, m_enableWriteBuffer))) {
-        log_error() << "Could not bind mount " << pathOnHost << " to " << dst;
+    // Bind mount to gateways
+    if (isError(bindMount(pathOnHost, tempDir, readonly, m_enableWriteBuffer))) {
+        log_error() << "Could not bind mount " << pathOnHost << " to " << tempDir;
         return ReturnCode::FAILURE;
     }
 
-    result = gatewaysDirInContainer() + "/" + pathInContainer;
+    std::string tempDirInContainer = gatewaysDirInContainer() + "/" + std::string(basename(pathInContainer.c_str()));
+
+    // Move the mount in the container
+    if (tempDirInContainer.compare(pathInContainer) != 0) {
+        pid_t pid = INVALID_PID;
+        ReturnCode mountMoveRes = executeInContainer([tempDirInContainer, pathInContainer] () {
+            unsigned long flags = MS_MOVE;
```

Comment:
Are you sure you don't want to keep the flags used before and just or with MS_MOVE?